### PR TITLE
feat(simulation): add bruteforce token slots detection

### PR DESCRIPTION
**Pull request by zizou0x**
Add a function that will try to brute force balance and allowance maps base slots for a given token. 

This function will iterate over the mapping "base_slot" and try to set a marker value to a particular key and read it. If the value is correctly retrieved, it means the "base_slot" is correct.

Also add a support for Vyper contracts for which the storage layout is different.

### CODE REVIEW

File: src/evm/mod.rs [hunk 1]
```diff
@@ -1,7 +1,48 @@
...
 pub mod engine_db_interface;
 pub mod simulation;
 pub mod simulation_db;
+pub mod token;
 pub mod traces;
 pub mod tycho_db;
 pub mod tycho_models;
+
+/// Enum representing the type of contract compiler.
+#[derive(Debug, PartialEq, Copy, Clone)]
```
Comment (zizou0x) [inline src/evm/mod.rs:15]: I made it `Copy` because it's a cheap enum

File: src/evm/token.rs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    static ref ERC20_ABI: Abi = {
+        let abi_file_path = "src/protocol/vm/assets/ERC20.abi";
+        let abi_json = std::fs::read_to_string(abi_file_path).expect("Failed to read ABI file");
+        from_str(&abi_json).expect("Failed to parse ABI JSON")
+    };
+}
+
+/// An error r
+#[derive(Error, Debug)]
+pub enum TokenError {
```
Comment (zizou0x) [inline src/evm/token.rs:29]: Is it overkill? I can make it a `String` but I like to have specific types for errors. Or if you have any suggestion about where this error could be added?

Comment (tamaralipows) [reply src/evm/token.rs:29]: I think this has to be a `SimulationError`

File: src/protocol/vm/state.rs [hunk 3]
```diff
@@ -522,6 +536,24 @@ impl VMPoolState<PreCachedDB> {
...
+                self.token_storage_slots.insert(
+                    *t,
+                    token::brute_force_slots(
+                        t,
+                        &self.block,
+                        self.engine
+                            .as_ref()
+                            .expect("engine should be set"),
+                    )
+                    .unwrap_or_else(|_| panic!("Couldn't brute force slots for {}", t)),
```
Comment (tamaralipows) [inline src/protocol/vm/state.rs:552]: I think we should return the error here like we do in the rest of the file.

**Review by tamaralipows — APPROVED** [review]
Thank you! Looks great

**Pull request merged**